impl/NotSoBasicLinearAlgebra.h

```diff
@@ -273,4 +273,104 @@ typename MemT::elem_t Trace(const Matrix<rows, cols, MemT> &A)
     return sum_diag;
 }
 
+
+template<int rows, int cols, class MemT = Array<rows, cols, float>> 
+class VectorValuedFunction{
+    private:
+        Matrix<rows, 1, Array<rows, 1, typename MemT::elem_t>> (*_f)(Matrix<cols, 1, Array<cols,1, typename MemT::elem_t>> x);
+
+    public:
+        VectorValuedFunction(){}
+
+        VectorValuedFunction(Matrix<rows, 1, Array<rows, 1, typename MemT::elem_t>> (&f)(Matrix<cols, 1, Array<cols,1, typename MemT::elem_t>> x)){_f = f;}
+        VectorValuedFunction(const VectorValuedFunction &V){_f = V._f;}
+        
+        ~VectorValuedFunction(){}
+
+        virtual Matrix<rows, 1, Array<rows, 1, typename MemT::elem_t>> vv_f(Matrix<cols, 1, Array<cols,1, typename MemT::elem_t>> x){
+            if(_f == nullptr){
+                return Matrix<rows, 1, Array<rows, 1, typename MemT::elem_t>>();
+            }else{
+                return _f(x);
+            }
+        }
+
+        Matrix<rows, 1, Array<rows, 1, typename MemT::elem_t>> operator ()(Matrix<cols, 1, Array<cols,1, typename MemT::elem_t>> x){
+            return vv_f(x);
+        }
+};
+
+template<int rows, int cols, class MemT = Array<rows, cols, float>>
+using VVF = VectorValuedFunction<rows, cols, MemT>;
+
+
+/**
+ * @brief numerically approximate the jacobian of a vector valued function. Cannot be used at the boundary of the domain
+ * 
+ * @tparam n inputs
+ * @tparam m outputs
+ * @tparam MemT 
+ * @param f vector valued functor 
+ * @param _x Point to take a derivative at
+ * @param h step size. Set at 0.0001, but for more precise applications, it can be lowered
+ * @return Matrix<n, m, MemT> 
+ */
+template <int n, int m, class MemT = Array<n, m, float> >
+Matrix<n, m, MemT> Jacobian(VectorValuedFunction<n,m,MemT> f, Matrix<m, 1, Array<m,1, typename MemT::elem_t>> _x,typename MemT::elem_t h = 0.0001)
+{  
+    Matrix<n, 1, Array<n, 1, typename MemT::elem_t>> f_x = f.vv_f(_x);
+
+    Matrix<n, m, MemT> Jacob;
+  
+    for(int i = 0; i < m; i++){
+        Matrix<m, 1, Array<m, 1, typename MemT::elem_t>> h_vec = Zeros<m>();
+        h_vec(i) = 1.00;
+        h_vec = h_vec * h;
+
+        Matrix<n, 1, Array<n, 1, typename MemT::elem_t>> f_xh = f.vv_f(_x + h_vec);
+
+        Matrix<n, 1, Array<n, 1, typename MemT::elem_t>> df = (f_xh - f_x)/h;
+
+        // Matrix<n,1,Array<n,1,typename MemT::elem_t>> df_mat = (Matrix<n,1,Array<n,1,typename MemT::elem_t>>) df;
+        
+        for(int j = 0; j < n; j++){
+            Jacob(j,i) = df(j);
+        }
+    }
+    return Jacob;
+}
+
+/**
+ * @brief get the gradient as a column vector
+ * 
+ * @tparam m 
+ * @tparam MemT 
+ * @param f 
+ * @param _x 
+ * @param h 
+ * @return Matrix<m, 1, MemT> 
+ */
+template<int m, class MemT = Array<m, 1, float>>
+Matrix<m, 1, MemT> Gradient(typename MemT::elem_t (*f)(Matrix<m, 1, Array<m,1, typename MemT::elem_t>> x), Matrix<m, 1, Array<m,1, typename MemT::elem_t>> _x,typename MemT::elem_t h = 0.0001){
```

Comment:
Since the gradient is a special case of the jacobian do you think it's worth defining a separate function for this case? We could get the same result by passing a function that returns a 1x1 `Matrix` to the `Jacobian` function right?